test/pylib/s3_server_mock.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/python3
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+# Mock S3 server to inject errors for testing.
+
+import os
+import sys
+import threading
+from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
+from urllib.parse import urlparse, parse_qs
+import uuid
+from enum import Enum
+from functools import partial
+from collections import OrderedDict
+
+sys.path.insert(0, os.path.dirname(__file__))
+
+
+class Policy(Enum):
+    SUCCESS = 0
+    RETRYABLE_FAILURE = 1
+    NONRETRYABLE_FAILURE = 2
+
+
+class LRUCache:
+    lock = threading.Lock()
+
+    def __init__(self, capacity: int):
+        self.cache = OrderedDict()
+        self.capacity = capacity
+
+    def get(self, key: str) -> Policy:
+        with self.lock:
+            if key not in self.cache:
+                return Policy.SUCCESS
+            self.cache.move_to_end(key)
+            return self.cache[key]
+
+    def put(self, key: str, value: Policy) -> None:
+        with self.lock:
+            if key in self.cache:
+                self.cache.move_to_end(key)
+            self.cache[key] = value
+            if len(self.cache) > self.capacity:
+                self.cache.popitem(last=False)
+
+
+class InjectingHandler(BaseHTTPRequestHandler):
```

Comment:
Need a comment explaining what this handler does. AFAIU it implements some state machine that tries to simulate multipart upload server and fail only the completion request, does it?

Reply:
Done